### DISCUSSION TITLE
A doubt regarding the type of contracts

### DISCUSSION THREAD

**Discussion by Jaseempk** [Cyfrin/foundry-full-course-cu#874] [Q&A]
Hey y'all,
which type of smart contracts is best?
Would you rather choose monolithic smart contract or modular smart contracts and what are your reasons in both cases?

Also would like top know which is often used in production, would like to hear some experienced advice on this .
Please help me out guys on this...

@PatrickAlphaC would be great to hear your take on this.

**Comment by Niferu** [comment]
Hello @Jaseempk 

First of all your question is wrong as there is no such things like mono or modu contracts for my knowledge at least. There are monolithic and modular blockchains. Here we are learning about ethereum blockchain only, so I believe you should redirect your question somewhere else.

There is also tons of information regarding this in google for example here: [link](https://www.alchemy.com/overviews/modular-vs-monolithic-blockchains)

That link should answer all of your questions

**Comment by Jaseempk** [comment]
@Neftyr Yea even I came to know that terms very recently, but its there I have read some articles on that but not much articles on this and even Claude2 & GPT did gave me examples for modular-contracts.
And the link you shared, yea that I have already came across while searching about this.

**Reply by Niferu** [reply]
@Jaseempk 

Smart contract will be modular because of blockchain is modular, so there is no point of dividing contracts because it depends on which blockchain those contracts are.

**Comment by Jaseempk** [comment]
<img width="1512" alt="Screenshot 2023-09-15 at 9 30 05 PM" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/99121494/14cdba7e-32e0-4161-ab94-b0aee354f382">

**Comment by theirrationalone** [comment]
Why not they don't exist??

In context of blockchains,

L1s are mono i.e., ETH, BTC. 
L2s are modular.

In context of smart contracts,

A single smart contract which has its all implementation in a single file with everything such as flexibility, security, rules, terms, orientation, scope, etc, are called mono. Examples would be, ERC20, ERC721 SMCs Or your any SMC with no deps. - one protocol Army.

Modular simply means, Main SMC imports depends on some outsourced Or external modules. Now you might got all the movie under the hood.

What to choose?? 

If i have a single one protocol army... It would be hard to abuse its security, hard to breach its rules, must to rely on implemented protocol. But if once got abused in some way, Your protocol will be a crap, therefore you will get burned, insolvent. 

Modular approach is the most famous approach even in web2 too. 

More modules more kind of layers with context focused security & rules on a specific implementation. But comes with more security holes, more ways to abuse rules & protocol. 


That's why, Internal & External keywords are designed and implemented into complier and made Available to us. 


The term monolithic is often used for addressing an independent, person, thing, place, philosophy, psychology. 

The term modular is often used for addressing a dependent, person, thing, place, philosophy, psychology.

Computer Science is the Psychology of statistics afterall. 

If i would to choose? I shall choose modular approach. Will make 100+ audits and will release protocol. One big adv. Is hard to breach protocol to noobs. Another big adv. Is, So much time consuming even to a pro too. 
.
